apps/src/pythonlab/PythonlabView.tsx

```diff
@@ -169,13 +173,41 @@ const PythonlabView: React.FunctionComponent<
   // Set view code to false if level is switched for any levels in widget view.
   useLifecycleNotifier(LifecycleEvent.LevelLoadStarted, () => {
     dispatch(setWidgetViewShowCode(false));
+    setAiTutor2HintQuestion(undefined);
   });
 
+  // Given a question for the AITutor2, return the full question to ask, which means
+  // appending all the relevant context.
+  const getAiTutor2FullQuestionFromQuestion = (
```

Comment:
could this be extracted out to a helper file? Also will this cause a re-render when the source is changed?

Reply:
> could this be extracted out to a helper file? 

Done.  Open to feedback on the first attempt!

> Also will this cause a re-render when the source is changed?

Good question.  Could you elaborate?

Reply:
Ah, interesting.  An underlying issue seems to be that a change in any variable in a context causes a re-render in any component that uses the context, regardless of which variables it actually depends upon.  There have been attempts such as [this](https://github.com/dai-shi/use-context-selector) to handle context variable changes in a more targeted way, but the mainstream solution appears to be to split contexts.  It seems we might have a broader opportunity to reduce re-renders outside of this PR; it seems that components like the Validation Results are already being re-rendered many times each time the source is clicked or typed into.

All that said, @sanchitmalhotra126 has some new work that builds on this PR which might hopefully reduce re-renders for the AI Tutor in particular...